scripts/staging/unified-memory-manager/umm.md

```diff
@@ -0,0 +1,144 @@
+provided by the JVM.
+
+The UMM has a certain (% of JVM heap) amount of memory which it can distribute for operands
+and a buffer pool.
+
+Operands are the compiled programs variables which extends the base class 
+`Data`, e.g., `MatrixObject` and `StringObject`.
+The buffer pool manages
+in memory representation of dirty objects that don't exist on the HDFS or 
+other persistent memory. This is currently done by the `LazyWriteBuffer`. Intermediates
+could be represented in this buffer.
+
+These two memory areas each will have a min and max amount of memory it can
+occupy, meaning that the boundary for the areas can shift dynamically depending
+on the current load.
+
+||min|max|
+| ------------- |:-------------:| -----:|
+| operations  | 50% | 70% |
+| buffer pool | 15% | 30% |
+
+The UMM will utilise the existing `CacheBlock` structure to manage the buffer
+pool area while it will use the new `OperandBlock (?)` structure to keep track of
+operations.
+
+### UMM Java Class
+For starters, we want to keep the API rather simple.
+
+```java
+/**
+ * Pins the input block into main memory.
+ * The provided block needs to fit into UMM boundaries.
+ */
+void pin(String, Block);
+
+/**
+ * Unpins the input block from main memory.
+ */
+void unpin(String);
+
+/**
+ * Loads the block into memory and pins it, if not evicted just pins it(?)
+ * The requested block needs to fit into UMM boundaries.
+ */
+Block serve(String);
+
+/**
+ * Reserves an specified amount in one of the two memory areas.
+ * The requested block needs to fit into UMM boundaries.
+ */
+Block reserve(Size, MemoryArea);
+
+/**
+ * Resizes reserved memory in one of the two memory areas.
+ * The requested block needs to fit into UMM boundaries.
+ */
+Block resize_reserved(Size, MemoryArea);
+
+```
+
+The UMM will also needs to keep track of the current state for which it will 
+have a few member variables. A queue similar to the current `EvictionQueue` is used
+to add/remove entries but rather than having a FIFO eviction policy 
+this queue will use LRU.
+
+
+```java
+class UMM {
+    // maximum capacity of the UMM
+    long _capacity;
+    // current operand area size
+    long _opSize;
+    // current buffer area size
+    long _bufferSize;
+    // operand block queue
+    EvictionQueue _opQueue;
+    // buffer block queue
+    EvictionQueue _bufferQueue;
+    // block handle, maybe a LocalVariableMap(?)
+    Map<String, Block> pinnedBlocks;
+}
+```
+
+
+### Block Java Class
+The first implementation will only handle CacheBlock objects, at the moment
+without any major changes to the CacheBlock object itself.
+
+If there is enough time this will be extended to an additional block type, `OperandBlock`.
+This would either require a new data type or maybe we could reuse some of the ideas
+from the `CacheableData<T extends CacheBlock> extends Data` class.
+
+### Testing
+Testing will be done with the existing java based test system which runs a
```

Comment:
For this exploratory project, I would recommend to only implement the memory manager, and add component tests that uses certain sequences of interactions and checks the final state. That way you can instantiate the memory manager with a small memory budget and test everything with tiny objects.